src/v/cloud_storage/partition_manifest.cc

```diff
@@ -305,6 +307,340 @@ partition_manifest::difference(const partition_manifest& remote_set) const {
+                _revision_id.value_or(model::initial_revision_id())),
+              .archiver_term = _archiver_term.value_or(model::term_id{})};
+            if (!_segments) {
+                _segments = std::unique_ptr<segment_map>(new segment_map());
+            }
+            _segments->insert(std::make_pair(_segment_key, _meta));
+            clear_meta_fields();
+            _state = state::expect_segment_path;
+            return true;
+        case state::expect_manifest_start:
+        case state::expect_manifest_value:
+        case state::expect_segments_start:
+        case state::expect_segment_meta_start:
+        case state::expect_segment_meta_value:
+        case state::terminal_state:
+            return false;
+        }
+    }
+
+    bool Bool(bool b) {
+        switch (_state) {
+        case state::expect_segment_meta_value:
+            if ("is_compacted" == _segment_meta_key) {
+                _is_compacted = b;
+                _state = state::expect_segment_meta_key;
+                return true;
+            }
+            return false;
+        case state::expect_manifest_start:
+        case state::expect_manifest_key:
+        case state::expect_manifest_value:
+        case state::expect_segments_start:
+        case state::expect_segment_path:
+        case state::expect_segment_meta_start:
+        case state::expect_segment_meta_key:
+        case state::terminal_state:
+            return false;
+        }
+    }
+
+    bool Null() {
+        switch (_state) {
+        case state::expect_manifest_value:
+            _state = state::expect_manifest_key;
+            return true;
+        case state::expect_segment_meta_value:
+            _state = state::expect_segment_meta_key;
+            return true;
+        case state::expect_manifest_start:
+        case state::expect_manifest_key:
+        case state::expect_segments_start:
+        case state::expect_segment_path:
+        case state::expect_segment_meta_start:
+        case state::expect_segment_meta_key:
+        case state::terminal_state:
+            return false;
+        }
+    }
+
+    bool Default() { return false; }
+
+    enum class state {
+        expect_manifest_start,
+        expect_manifest_key,
+        expect_manifest_value,
+        expect_segments_start,
+        expect_segment_path,
+        expect_segment_meta_start,
+        expect_segment_meta_key,
+        expect_segment_meta_value,
+        terminal_state,
+    } _state{state::expect_manifest_start};
+
+    using segment_map = partition_manifest::segment_map;
+
+    key_string _manifest_key;
+    key_string _segment_meta_key;
+    segment_name _segment_name;
+    std::optional<segment_name_components> _parsed_segment_key;
+    partition_manifest::key _segment_key;
+    partition_manifest::segment_meta _meta;
+    std::unique_ptr<segment_map> _segments;
+
+    // required manifest fields
+    std::optional<int32_t> _version;
+    std::optional<model::ns> _namespace{};
```

Comment:
nit: these don't need to be `{}` initialized--they are already default initialized.

Reply:
fixed